File: leetcode/cpp/206.reverse-linked-list.cpp
#include "oj_header.h"

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        if (!head || !head->next)
        {
            return head;
        }

        auto rest = reverseList(head->next); //反转后n-1个链表
        head->next->next = head;
        /// \note 反转后的链表的尾结点应当指向 `nullptr`
        head->next = nullptr;

        return rest;
    }
};

File: leetcode/rust/src/206.reverse-linked-list.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if head.is_none() {
            return head;
        }

        let mut  prev : Option<Box<ListNode>> = Option::None;
        let mut curr : Option<Box<ListNode>> = head;
        while  curr.is_some() {
            let mut curr_box = curr.unwrap();
            let next = curr_box.next.take();

            curr_box.next = prev;
            prev = Some(curr_box);
            curr = next;
        }

        prev
    }
}